### DISCUSSION TITLE
some basics for someone coming from Django

### DISCUSSION THREAD

**Discussion by developer992** [sqlalchemy/sqlalchemy#10916] [Usage Questions]
Here's what I need

I need two or three tables. I would like to have them defined in a models.py file, so I can keep it under version control.
Secondly, i need migrations .. trying to use alembic
I have PostgreSQL database server

- how do i define a table ... googling this, i found several ways to define a table but very little information which one is prefferred (https://docs.sqlalchemy.org/en/20/core/metadata.html vs https://stackoverflow.com/questions/45259764/how-to-create-a-single-table-using-sqlalchemy-declarative-base vs )
- what even is declarative base and metadata? 
- how do i keep DATABASE_URL in my env file and not in my alembic.ini file ?
- how do init async engine

Model:

```
from sqlalchemy import Table, Column, Integer, String, MetaData, Uuid, DateTime
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.ext.asyncio import AsyncAttrs


class Base(AsyncAttrs, DeclarativeBase):
    pass


class ATable(Base):
    __tablename__ = "my_table_name"

    id = mapped_column(Integer, primary_key=True)

    # Title
    title = mapped_column(String(255))

    # ... other fields
```

then in my env.py file:
I import above Base
and do this:
```target_metadata = Base.metadata```

Where does DATABASE_URL come into the equation?

Alembic autogenerate is failing
```alembic revision --autogenerate -m "Initial"```

```
  File ...\alembic\util\pyfiles.py", line 113, in load_module_py
    spec.loader.exec_module(module)  # type: ignore
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "...\alembic\env.py", line 81, in <module>
    run_migrations_online()
  File "...\alembic\env.py", line 63, in run_migrations_online
    connectable = engine_from_config(
                  ^^^^^^^^^^^^^^^^^^^
  File "...\sqlalchemy\engine\create.py", line 810, in engine_from_config
    url = options.pop("url")
          ^^^^^^^^^^^^^^^^^^
KeyError: 'url'

```

It loloks like these two generated methods aren't correct
env.py::run_migrations_online and
env.py::run_migrations_offline

it is trying to initialize engine from config but failing because URL

**Comment by developer992** [comment]
ok apparently, DeclarativeBase is newer, in sqlalchemy 2.0 and is preferred ( source https://docs.sqlalchemy.org/en/20/changelog/whatsnew_20.html#migrating-an-existing-mapping )

buit every guide i find for this is wrong or old, not sure which

https://python.plainenglish.io/understanding-sqlalchemys-declarative-base-and-metadata-in-python-8eb0946b0eff
```

    metadata = MetaData(bind=engine)
               ^^^^^^^^^^^^^^^^^^^^^
TypeError: MetaData.__init__() got an unexpected keyword argument 'bind'
```

**Comment by developer992** [comment]
i do feel like following the tutorial should yield results and not just errors

**Reply by zzzeek** [reply]
follow the correct tutorial will help

https://docs.sqlalchemy.org/en/20/tutorial/index.html

**Reply by developer992** [reply]
does this one help with async engine?

**Comment by developer992** [comment]
so why is here then different syntax? https://docs.sqlalchemy.org/en/20/tutorial/metadata.html

**Reply by zzzeek** [reply]
what, that `MetaData(bind=XYZ)` doesn't work?  that's an old version of SQLAlchemy, that's not how it looks anymore

**Reply by developer992** [reply]
so where does it say this and how should i know this?

**Reply by developer992** [reply]
its also under the 2.0 url so for which version is this?

i think you guys need to rewrite your docs

**Reply by zzzeek** [reply]
why would you know about "bind" at all?  Read only SQLAlchemy's docs, you wont see such a thing, then you wont be confused.  if you read out of date tutorials somewhere else, assume they are wrong.

**Reply by zzzeek** [reply]
already did!  what do you think ?

**Comment by developer992** [comment]
i think i dont know where to start anymore ...

**Comment by developer992** [comment]
so where in here is syntax for defining a table?
https://docs.sqlalchemy.org/en/20/tutorial/index.html

**Comment by developer992** [comment]
so much about that great tutorial innit

**Reply by zzzeek** [reply]
OK we are done here.  if you have real questions and discussions to have, open a new discussion thanks
